### PR TITLE
Increase reliability of communication with Planet servers and avoid clobbering them

### LINKED ISSUE #551: Handle additional rate-limit failure modes with exponential backoff.
#279 addressed handling HTTP 429 TOO_MANY_REQUESTS failures through use of retry with exponential backoff.

Use of the Data and Orders APIs have shown additional failure modes in rate-limiting situations beyond just the HTTP 429 response, that should be handled in the same way. These include timeouts and partial responses, which may occur when many (1000+) API requests are submitted all at once, and the scaling and rate-limiting mechanisms of the APIs fail to catch them all, and may manifest as the following exceptions (assuming use of the requests library):

- requests.exceptions.ChunkedEncodingError
- requests.exceptions.ReadTimeout

In addition to these failures, occasionally 4xx or 5xx responses are returned with a blank or corrupt body, such that attempting to parse the response body results in some form of decode error (e.g. `json.decoder.JSONDecodeError` when calling `response.json()`) - this appears to be another manifestation of the above failure.

I have also observed intermittent failures specifically when submitting orders, where I receive a 5xx status with the message "Problem confirming products". This failure may also be related to rate-limiting, as a resubmission of the same order always appears to succeed. I am less certain as to whether this particular failure should be handled through the same method, as there may be cases where it is actually a relevant response.

### PR DESCRIPTION
**Pull request by jreiberkyle**
This introduces three major changes to communication with the Planet services to increase reliability of the communication and avoid clobbering the services.

1. Every request sent by a Session is now limited according to maximum workers and rate_limit, imposed by the new `_Limiter` class.
2. Increases the set of exceptions that trigger a (exponential-backoff) retry.
3. Increases read timeout.

Additionally, this adds the script that was used to identify, explore, and test the configuration and populate the exception list. 

Results from testing and findings that led to configuration settings and retry exception identification can be found in #580. 

This also closes #551 and #578.

TODO checklist:

- [x] retest final settings with script
- [x] document in http how settings were obtained and how to rerun script to tweak them if they aren't working for another endpoint

### CODE REVIEW

File: tests/unit/test_http.py [hunk 1]
```diff
@@ -71,6 +71,30 @@ def test_basesession__raise_for_status(mock_response):
         http.BaseSession._raise_for_status(
             mock_response(HTTPStatus.METHOD_NOT_ALLOWED, json={}))
 
+import asyncio
+import time
+@pytest.mark.asyncio
+async def test__Limiter_rate_limit():
```
Comment (jreiberkyle) [inline tests/unit/test_http.py:77]: @sgillies any ideas for how we can test the rate/worker limiting? Its a tricky async/timing problem

File: planet/http.py [hunk 2]
```diff
@@ -15,19 +15,39 @@
...
+import httpcore
 import httpx
 
 from .auth import Auth, AuthType
 from . import exceptions, models
 from .__version__ import __version__
 
+# For how this list was determined, see
+# https://github.com/planetlabs/planet-client-python/issues/580
+RETRY_EXCEPTIONS = [
```
Comment (jreiberkyle) [inline planet/http.py:33]: @cholmes see extra exceptions caught here

File: planet/http.py [hunk 3]
```diff
@@ -99,10 +99,27 @@ def _raise_for_status(cls, response):
...
+    close to eachother (even though max rate limit is 5 calls per second)[2].
+
+    In investigating options, aiometer[3] was also looked at but it seems to
+    have odd behavior with httpx [4].
+
+    References:
+    [1] https://github.com/mjpieters/aiolimiter
+    [2] https://github.com/planetlabs/planet-client-python/issues/580#issuecomment-1182752851
+    [3] https://github.com/florimondmanca/aiometer
+    [4] https://github.com/florimondmanca/aiometer/issues/24
```
Comment (sgillies) [inline planet/http.py:120]: References and design notes are great! This is a good example for projects to follow.

File: tests/unit/test_http.py (hunk 1 above)
Comment (sgillies) [reply tests/unit/test_http.py:77]: You mean testing the cadence? How about wrapping the limiter around calls to time.monotonic, collecting N number of times, and then asserting that the N calls don't exceed the Limiters rate limit and that no calls are closer together than the cadence threshold? The limiter can throttle anything, not only HTTP requests, right? 

```python
async def get_time_limited(limiter):
    async with limiter:
        return time.monotonic()

limiter = Limiter(...)
times = [get_time_limited(limiter) async for i in range(100)]
assert all(second - first > threshold for first, second in itertools.pairwise(times))
```

Is that useful?

Comment (jreiberkyle) [reply tests/unit/test_http.py]: yeah, it can limit anything! This is a cool approach and is helpful.

File: planet/http.py (hunk 3 above)
Comment (jreiberkyle) [reply planet/http.py]: thanks! I agree wholeheartedly

**Comment by jreiberkyle** [comment]
@sgillies @mkshah605 @kevinlacaille and @tbarsballe FYI.

**Pull request merged**